### PR TITLE
function correlation_lags is added

### PR DESCRIPTION
**Pull request by Vidhi-44**
This is my first PR . A function correlation_lags is added . Waiting for the review . I will be adding test cases in a day.

### CODE REVIEW

File: cupyx/scipy/signal/_signaltools.py [hunk 1]
```diff
@@ -411,6 +411,40 @@ def _correlate2d(in1, in2, mode, boundary, fillvalue, convolution=False):
     return _st_core._direct_correlate(in1, in2, mode, in1.dtype, convolution,
                                       boundary, fillvalue, not convolution)
 
+def correlation_lags(in1_len, in2_len, mode='full'):
+    # calculate lag ranges in different modes of operation
```
Comment (emcastillo) [inline cupyx/scipy/signal/_signaltools.py:415]: Please add a docstring for docs generation

**Review by emcastillo — CHANGES_REQUESTED** [review]
Thanks for the PR! Can you add tests?

File: cupyx/scipy/signal/_signaltools.py (hunk 1 above)
Comment (Vidhi-44) [reply cupyx/scipy/signal/_signaltools.py:415]: Docstring has been added.

File: cupyx/scipy/signal/_signaltools.py [hunk 2]
```diff
@@ -411,6 +410,101 @@ def _correlate2d(in1, in2, mode, boundary, fillvalue, convolution=False):
...
+        \overline{f\left [ m \right ]}g\left [ m+n \right ]
+
+    Where :math:`n` is the lag.
+
+    Examples
+    --------
+    Cross-correlation of a signal with its time-delayed self.
+
+    >>> import cupy
+    >>> from scipy import signal
```
Comment (emcastillo) [inline cupyx/scipy/signal/_signaltools.py:466]:
```suggestion
    >>> from cupyx.scipy import signal
```

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 3]
```diff
@@ -207,6 +207,31 @@ def test_convolve2d(self, xp, scp, dtype):
                                  accept_error=ValueError)
     def test_correlate2d(self, xp, scp, dtype):
         return self._filter('correlate2d', dtype, xp, scp)
+       
+    
+    def test_correlation_lags(mode, behind, input_size):
```
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:212]:
```suggestion
    @testing.numpy_cupy_allclose(atol=tols, rtol=tols, scipy_name='scp',
                                                          accept_error=ValueError)
    def test_correlation_lags(mode, behind, input_size):
```

File: cupyx/scipy/signal/_signaltools.py (hunk 2 above)
Comment (Vidhi-44) [reply cupyx/scipy/signal/_signaltools.py]: Hey @emcastillo . 
Changes made. Waiting for your review.

**Comment by emcastillo** [comment]
please, rewrite the test like the ones above it. 
Also verify that it works locally with pytest before submitting it. Thanks

**Comment by Vidhi-44** [comment]
> please, rewrite the test like the ones above it. Also verify that it works locally with pytest before submitting it. Thanks

Sorry, I couldn't understand what do you mean by this:
"Rewrite the test like the ones above it". Please specify what you are referring to @emcastillo

**Comment by emcastillo** [comment]
Sorry, can you fix the style errors that the CI reports? thanks

**Comment by Vidhi-44** [comment]
Hey @emcastillo .
I have been trying to figure out the style errors for which I had installed pre-commit and then made few changes but I was not able to get passed all the files which were failing so I tried to push the code may be the pre-commit was not allowing it due to the failed ones. 
Please help me out .




[WARNING] The 'files' field in hook 'ci-generate' is a regex, not a glob -- matching '/*' probably isn't what you want here
[INFO] Installing environment for https://github.com/pre-commit/mirrors-autopep8.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
check for added large files..............................................Passed
ensure no direct commit to master/vXX branch.............................Passed
check for case conflicts.................................................Passed
mixed line ending........................................................Passed
fix utf-8 byte order marker..............................................Passed
check yaml...............................................................Passed
autopep8.................................................................Failed
- hook id: autopep8
- files were modified by this hook
flake8...................................................................Failed
- hook id: flake8
- exit code: 1

cupyx/scipy/signal/_signaltools.py:447:9: W605 invalid escape sequence '\l'
cupyx/scipy/signal/_signaltools.py:447:18: W605 invalid escape sequence '\s'
cupyx/scipy/signal/_signaltools.py:447:34: W605 invalid escape sequence '\l'
cupyx/scipy/signal/_signaltools.py:448:20: W605 invalid escape sequence '\i'
cupyx/scipy/signal/_signaltools.py:449:9: W605 invalid escape sequence '\o'
cupyx/scipy/signal/_signaltools.py:449:20: W605 invalid escape sequence '\l'
cupyx/scipy/signal/_signaltools.py:449:40: W605 invalid escape sequence '\l'
cupyx/scipy/signal/_signaltools.py:457:9: W605 invalid escape sequence '\l'
cupyx/scipy/signal/_signaltools.py:457:18: W605 invalid escape sequence '\s'
cupyx/scipy/signal/_signaltools.py:457:34: W605 invalid escape sequence '\l'
cupyx/scipy/signal/_signaltools.py:458:20: W605 invalid escape sequence '\s'
cupyx/scipy/signal/_signaltools.py:458:27: W605 invalid escape sequence '\i'
cupyx/scipy/signal/_signaltools.py:458:36: W605 invalid escape sequence '\i'
cupyx/scipy/signal/_signaltools.py:459:9: W605 invalid escape sequence '\o'
cupyx/scipy/signal/_signaltools.py:459:20: W605 invalid escape sequence '\l'
cupyx/scipy/signal/_signaltools.py:459:40: W605 invalid escape sequence '\l'

flake8 (cython)......................................(no files to check)Skipped
mypy.....................................................................Passed
mypy (install).......................................(no files to check)Skipped
check CI files are up-to-date (.pfnci/generate.py)...(no files to check)Skipped

File: cupyx/scipy/signal/_signaltools.py [hunk 4]
```diff
@@ -411,6 +410,101 @@ def _correlate2d(in1, in2, mode, boundary, fillvalue, convolution=False):
     return _st_core._direct_correlate(in1, in2, mode, in1.dtype, convolution,
                                       boundary, fillvalue, not convolution)
 
+def correlation_lags(in1_len, in2_len, mode='full'):
```
Comment (emcastillo) [inline cupyx/scipy/signal/_signaltools.py:413]:
```suggestion

def correlation_lags(in1_len, in2_len, mode='full'):
```

File: cupyx/scipy/signal/_signaltools.py [hunk 5]
```diff
@@ -411,6 +410,101 @@ def _correlate2d(in1, in2, mode, boundary, fillvalue, convolution=False):
...
+        # must be at least as large as the other in every dimension.
+
+        # the lag_bound will be either negative or positive
+        # this let's us infer how to present the lag range
+        lag_bound = in1_len - in2_len
+        if lag_bound >= 0:
+            lags = cupy.arange(lag_bound + 1)
+        else:
+            lags = cupy.arange(lag_bound, 1)
+    return lags
```
Comment (emcastillo) [inline cupyx/scipy/signal/_signaltools.py:507]:
```suggestion
    return lags

```

File: cupyx/scipy/signal/_signaltools.py [hunk 6]
```diff
@@ -7,7 +7,6 @@
 from cupyx.scipy.ndimage import _filters
 from cupyx.scipy.signal import _signaltools_core as _st_core
 
-
 def convolve(in1, in2, mode='full', method='auto'):
```
Comment (emcastillo) [inline cupyx/scipy/signal/_signaltools.py:10]:
```suggestion

def convolve(in1, in2, mode='full', method='auto'):
```

File: cupyx/scipy/signal/_signaltools.py [hunk 7]
```diff
@@ -412,6 +412,103 @@ def _correlate2d(in1, in2, mode, boundary, fillvalue, convolution=False):
...
+
+    Cross correlation for discrete functions :math:`f` and :math:`g` is
+    defined as:
+
+    .. math::
+        \left ( f\star g \right )\left [ n \right ]
+        \triangleq \sum_{-\infty}^{\infty}
+        \overline{f\left [ m \right ]}g\left [ m+n \right ]
+
+    Where :math:`n` is the lag.
```
Comment (emcastillo) [inline cupyx/scipy/signal/_signaltools.py:461]:
```suggestion
```

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 8]
```diff
@@ -208,6 +209,36 @@ def test_convolve2d(self, xp, scp, dtype):
...
 
+    def test_correlation_lags(mode, behind, input_size):
+        # generate random data
+        rng = cupy.random.RandomState(0)
+        in1 = rng.standard_normal(input_size)
+        offset = int(input_size/10)
+        # generate offset version of array to correlate with
+        if behind:
+            # y is behind x
+            in2 = cupy.concatenate([rng.standard_normal(offset), in1])
```
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:220]:
```suggestion
            in2 = xp.concatenate([rng.standard_normal(offset), in1])
```

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 9]
```diff
@@ -208,6 +209,36 @@ def test_convolve2d(self, xp, scp, dtype):
...
+            expected = -offset
+        else:
+            # y is ahead of x
+            in2 = in1[offset:]
+            expected = offset
+        # cross correlate, returning lag information
+        correlation = correlate(in1, in2, mode=mode)
```
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:227]:
```suggestion
        correlation = scp.correlate(in1, in2, mode=mode)
```

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 10]
```diff
@@ -208,6 +209,36 @@ def test_convolve2d(self, xp, scp, dtype):
...
+        lags = correlation_lags(in1.size, in2.size, mode=mode)
```
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:228]:
```suggestion
        lags = scp.signal.correlation_lags(in1.size, in2.size, mode=mode)
```

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 11]
```diff
@@ -208,6 +209,36 @@ def test_convolve2d(self, xp, scp, dtype):
...
+        # identify the peak
+        lag_index = cupy.argmax(correlation)
```
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:230]:
```suggestion
        lag_index = xp.argmax(correlation)
```

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 12]
```diff
@@ -208,6 +209,36 @@ def test_convolve2d(self, xp, scp, dtype):
     def test_correlate2d(self, xp, scp, dtype):
         return self._filter('correlate2d', dtype, xp, scp)
 
+    def test_correlation_lags(mode, behind, input_size):
```
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:212]:
```suggestion
    def test_correlation_lags(self, mode, behind, input_size):
```

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 13]
```diff
@@ -208,6 +209,36 @@ def test_convolve2d(self, xp, scp, dtype):
...
+        # generate random data
+        rng = cupy.random.RandomState(0)
+        in1 = rng.standard_normal(input_size)
+        offset = int(input_size/10)
```
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:216]:
```suggestion
        offset = int(self.input_size/10)
```

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py (hunk 9 above)
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:227]:
```suggestion
        correlation = correlate(in1, in2, mode=self.mode)
```

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 14]
```diff
@@ -208,6 +209,36 @@ def test_convolve2d(self, xp, scp, dtype):
...
+        # cross correlate, returning lag information
+        correlation = correlate(in1, in2, mode=mode)
+        lags = correlation_lags(in1.size, in2.size, mode=mode)
+        # identify the peak
+        lag_index = cupy.argmax(correlation)
+        # Check as expected
+        assert_equal(lags[lag_index], expected)
+        # Correlation and lags shape should match
+        assert_equal(lags.shape, correlation.shape)
+
```
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:235]:
```suggestion
          return lags  
```

File: tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py [hunk 15]
```diff
@@ -208,6 +209,36 @@ def test_convolve2d(self, xp, scp, dtype):
...
+            expected = offset
+        # cross correlate, returning lag information
+        correlation = correlate(in1, in2, mode=mode)
+        lags = correlation_lags(in1.size, in2.size, mode=mode)
+        # identify the peak
+        lag_index = cupy.argmax(correlation)
+        # Check as expected
+        assert_equal(lags[lag_index], expected)
+        # Correlation and lags shape should match
+        assert_equal(lags.shape, correlation.shape)
```
Comment (emcastillo) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_signaltools.py:234]:
```suggestion
        assert_equal(lags.shape, correlation.shape)
        return lags
```

**Comment by emcastillo** [comment]
test your code locally before submitting it to review please

**Comment by emcastillo** [comment]
Let's close this meanwhile :). feel free to reopen when you have updates!

**Pull request closed without merging**